New & improved R package installer for repgen.

ashalper-usgs opened the pull request.

ashalper-usgs committed:
devtools installation upgrade.

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Reset to FETCH_HEAD to synchronize with GitHub USGS-R/repgen master.

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Initial version.

ashalper-usgs committed:
Obsolete.

ashalper-usgs committed:
Steering towards configuration found on nwissddvasaqcu.

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Merge remote-tracking branch 'GitHub/master' into AQCU-949

ashalper-usgs committed:
Added conditional to remove bootstrapping development packages:

on servers that do not need them.

ashalper-usgs committed:
Added conditional to check R_LIBS environment variable.

ashalper-usgs committed:
Added "evaluate" to pkgs.

ashalper-usgs committed:
Re-purposed this file to contain common, installPackages().

ashalper-usgs committed:
The file formerly known as installPackages.R.

ashalper-usgs committed:
Added default for "repos" argument.

ashalper-usgs committed:
Initial version.

ashalper-usgs committed:
Removed some debugging code.

ashalper-usgs committed:
Added some smarts to:

source installPackages.R correctly in production environment.

ashalper-usgs committed:
Obsolete.

lindsayplatt commented on inst/extdata/dev-packages.R:
```diff
@@ -0,0 +1,6 @@
+
+source(paste0(getwd(), "/inst/extdata/installPackages.R"))
+
+installPackages(c("devtools", "roxygen2", "testthat"))
```
:+1:

mhines-usgs commented on inst/extdata/packages.R:
```diff
@@ -0,0 +1,94 @@
+
+# Known production tiers, listed at
+# https://docs.google.com/document/d/1vBOTUPtIdeCTGe7Or2cQGMMOFS_hD545yZ1E6YBGizs/edit
+tiers <-
```
I don't know that we shouldn't have the info here in the script but why couldn't this come from Jenkins?

thongsav-usgs replied:
Agreed, we should not be checking in infrastructure information into project code. The application should not know or react to infrastructure it's on.

thongsav-usgs replied:
If the desire is to have this parameterized, it could be a simple "includeDevPackages" boolean, and the caller would decide what to pass in. Versus the script code deciding based on where it's sitting.

ashalper-usgs replied:
@thongsav-usgs how would you have the code remove `BH`, `devtools`, `httr` from a production system? These are development packages that are only required to install `gsplot`. After `gsplot` is installed, they become cruft.

thongsav-usgs replied:
Have the caller of the script pass in the boolean saying to remove dev packages (either pass in a param or set an env variable), the application itself should not have tier specific logic.

ashalper-usgs replied:
OK, will do.

mhines-usgs commented on inst/extdata/installRDependencies.sh:
```diff
@@ -1,26 +0,0 @@
-
-Rscript -e 'install.packages("httr", repos = "http://mran.microsoft.com/snapshot/2016-01-27",type="source")'
-Rscript -e 'install.packages("devtools", repos = "http://mran.microsoft.com/snapshot/2015-01-19",type="source")'
-Rscript -e 'Sys.setenv(TZ="UTC")'
```
Is this no longer needed?

ashalper-usgs replied:
Correct. `packages.R`/`installPackages.R` supersedes it.

mhines-usgs replied:
I'm specifically asking about the timezone line here

-Rscript -e 'Sys.setenv(TZ="UTC")'

ashalper-usgs replied:
owe, good point. I wasn't sure about that. Should that go in `packages.R`?

mhines-usgs replied:
sure, i don't think it much matters as long as it happens as a part of the install/set up.  OR if you're doing any other environment-specific bits, maybe it makes sense to do this there? like where you set the R_LIBS? @thongsav-usgs is there a specific place this needs to be called at? I know we have done it 'last' but not sure it needs to be that way.

mhines-usgs commented on inst/extdata/packages.R:
```diff
@@ -0,0 +1,94 @@
+
+# all packages except devtools and its prerequisites are held back to older
+# versions, keyed by their repo URLs below
+installPackages(pkgs, "http://mran.microsoft.com/snapshot/2016-03-31")
```
So this also handles all the required packages upstream dependencies?

ashalper-usgs replied:
Yes. `install.packages()` does that by default.

mhines-usgs replied:
For some reason I remember explicitly calling the requirements in order but can't remember why that mattered, hopefully it no longer does.

ashalper-usgs replied:
I've stared at the diffs between DEV and my workstation quite a lot now. I suspect R's `install.packages()` is smarter now than when the original package installer was written.

Also it's crossed my mind that since we're in the process of doing this rather major re-factor, it might make sense to bring the package snapshot set up to the present date. I've noticed that the package dependency tree for the development packages at least, seems to have changed since then. I would expect the same for the production packages too.

mhines-usgs replied:
I think it's fine, as long as we are able to verify moving the mran date to a later date works as expected. But I think it makes sense to wait until we're done with the refactor and then change that date? I'd make a future task for it, I think.

ashalper-usgs replied:
If it were me (and it's mostly not), and I was being all Agile, I'd want the upstream prerequisites established earlier and oftener; so that I knew if there were any API changes in the newer package versions, I could cope with those earlier in repgen.

mhines-usgs replied:
I think we want to update but not in the middle of the refactor, is my point. I think others agreed in slack when we had this convo.

ashalper-usgs replied:
As you like. Mostly not me.

mhines-usgs replied:
I made this skeleton task that we should pull in when we're done with the refactor, feel free to add details or suggestions!

https://internal.cida.usgs.gov/jira/browse/AQCU-1009

ashalper-usgs committed:
Re-worked how development packages are managed.

ashalper-usgs committed:
Made error handling smarter.

ashalper-usgs commented:
Please don't merge this yet. Found a problem.

ashalper-usgs committed:
Merge branch 'master' of ssh://git@github.com/USGS-R/repgen.git

ashalper-usgs committed:
Merge remote-tracking branch 'GitHub/master' into AQCU-936

ashalper-usgs committed:
Moved R_LIBS check up to packages.R.

ashalper-usgs committed:
Gave up on command line argument passing for the moment.

ashalper-usgs committed:
Added "else" to install developer packages.

ashalper-usgs committed:
Obsolete.

ashalper-usgs committed:
Re-re-factored.

ashalper-usgs committed:
Obsolete.

ashalper-usgs committed:
Clean-up and bug fixes:

Cleaned up some comments. Removed now-infinitely-recursive source()
calls of installPackages.R (this file). Dumbed-down code that attempted
to install Rtools.

ashalper-usgs committed:
Added a TODO and removed call to .rs.restartR():

as it only exists in RStudio.

mhines-usgs commented:
@ashalper-usgs you asked me to review this again but there isn't anything new.

ashalper-usgs commented:
Kind of don't know what is going on. I closed this PR, and opened a new one: https://github.com/USGS-R/repgen/pull/555
